### ISSUE TITLE
java.lang.IllegalStateException: open when making GORM calls

### ISSUE THREAD

**Issue by amexboy** [fakemongo/fongo#89]
In my grails 2.3.11 project I wanted to use Fongo instead of mongo for integration tests. I replaced the bean called `mongo` created by mongoBD plugin to be an instance of Fongo. The plugin creates a bean called `mongo` which is an instance of `GMongo`, which contains an instance of `Mongo` of the mongoBD plugin. Therefor, to be able to tunnel GORM called to mongo, I replaced the bean to contain an instance of mongo. 

The bean I replaced works for other calls! E.g, this works 
```
mongo.getDB("terra_test").user.insert([name:"amanuel"])

```
but this how ever 
```
new User(username: "Nasdjfkl", password: "klajsdk;fj").save(flush: true, failOnError: true)
```
produces this stacktrace
```
java.lang.IllegalStateException: open
	at org.bson.util.Assertions.isTrue(Assertions.java:36)
	at com.mongodb.DBTCPConnector.isMongosConnection(DBTCPConnector.java:369)
	at com.mongodb.Mongo.isMongosConnection(Mongo.java:645)
	at com.mongodb.DBCursor._check(DBCursor.java:454)
	at com.mongodb.DBCursor._hasNext(DBCursor.java:546)
	at com.mongodb.DBCursor.hasNext(DBCursor.java:571)
	at org.grails.datastore.mapping.mongo.query.MongoQuery$MongoResultList.isEmpty(MongoQuery.java:1268)
	at org.grails.datastore.gorm.finders.AbstractFindByFinder.invokeQuery(AbstractFindByFinder.java:33)
	at org.grails.datastore.gorm.finders.AbstractFindByFinder$1.doInSession(AbstractFindByFinder.java:24)
	at org.grails.datastore.mapping.core.DatastoreUtils.execute(DatastoreUtils.java:302)
	at org.grails.datastore.gorm.finders.AbstractFinder.execute(AbstractFinder.java:41)
	at org.grails.datastore.gorm.finders.AbstractFindByFinder.doInvokeInternal(AbstractFindByFinder.java:22)
	at org.grails.datastore.gorm.finders.DynamicFinder.invoke(DynamicFinder.java:152)
	at org.grails.datastore.gorm.finders.DynamicFinder.invoke(DynamicFinder.java:352)
	at org.grails.datastore.gorm.GormStaticApi.methodMissing(GormStaticApi.groovy:105)
	at org.grails.datastore.gorm.internal.StaticMethodInvokingClosure.call(StaticMethodInvokingClosure.groovy:32)
	at org.grails.datastore.gorm.validation.constraints.UniqueConstraint.processValidate_closure1(UniqueConstraint.groovy:47)
	at org.grails.datastore.gorm.validation.constraints.UniqueConstraint.withManualFlushMode_closure2(UniqueConstraint.groovy:109)
	at org.grails.datastore.gorm.GormStaticApi.withSession_closure22(GormStaticApi.groovy:682)
	at org.grails.datastore.mapping.core.DatastoreUtils.execute(DatastoreUtils.java:302)
	at org.grails.datastore.gorm.AbstractDatastoreApi.execute(AbstractDatastoreApi.groovy:37)
	at org.grails.datastore.gorm.GormStaticApi.withSession(GormStaticApi.groovy:681)
	at org.grails.datastore.gorm.validation.constraints.UniqueConstraint.withManualFlushMode(UniqueConstraint.groovy:104)
	at org.grails.datastore.gorm.validation.constraints.UniqueConstraint.processValidate(UniqueConstraint.groovy:39)
	at org.grails.datastore.gorm.GormValidationApi.doValidate(GormValidationApi.groovy:64)
	at org.grails.datastore.gorm.GormValidationApi.validate(GormValidationApi.groovy:145)
	at org.grails.datastore.gorm.GormInstanceApi.doSave(GormInstanceApi.groovy:172)
	at org.grails.datastore.gorm.GormInstanceApi.save_closure5(GormInstanceApi.groovy:162)
	at org.grails.datastore.mapping.core.DatastoreUtils.execute(DatastoreUtils.java:302)
	at org.grails.datastore.gorm.AbstractDatastoreApi.execute(AbstractDatastoreApi.groovy:37)
	at org.grails.datastore.gorm.GormInstanceApi.save(GormInstanceApi.groovy:161)
	at com.apposit.terra.connect.service.integration.test.OrganizationUserServiceSpec.test db workd(OrganizationUserServiceSpec.groovy:143)
```

If i wrap the GORM call in a try catch and repeat the first statement, It still works.

**Comment by twillouer** [comment]
Hi,

thanks for the report.

I did not understand the stacktrace.
 com.mongodb.Mongo.isMongosConnection(Mongo.java:645) is normally used to know if you're connected to a "mongos" process.

But in your stacktrace, it's seems that you're not really in a fongo context.

Do you have a local database who is running in background ?
I did not see any references to fongo in the stackstrace (and isMongosConnection is overloaded in com.mongodb.MockMongoClient#isMongosConnection), so the result is strange

**Comment by amexboy** [comment]
Yes, I have! But it is not connected. mongoDB plugin creates a bean called mongo and that holds a connection to a mongo, which i overriden to be a connection to Fongo. The problem I faced (I know better now) Is gorm does not connect via this bean. There is another bean called `mongoDataStore`. I needed to override that too. My new problem now is Fongo does not have the alternative of this on! 

As far as my research went, the solution is to also override the `mongoDataStore` bean! I am working to create a plugin for grails that uses fongo as a data storage for really fast Integration tests. [Fongo-for-Integration-tests](https://github.com/amexsofts/fongo-for-integration-test) .

Basically, What I want to achieve is override every call to the mogo process to be to fongo. From what I see now, you have support for Unit tests only!
